src/eos/twoPhase.cpp

```diff
@@ -0,0 +1,1075 @@
+    auto functionContext = (FunctionContext *)ctx;
+
+    // get the velocity for kinetic energy
+    PetscReal density = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHO];
+    PetscReal ke = 0.0;
+    for (PetscInt d = 0; d < functionContext->dim; d++){
+        ke += PetscSqr(conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOU + d] / density);
+    }
+    ke *= 0.5;
+
+    // compute internal energy
+    *internalEnergy = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOE] / density - ke;
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::eos::TwoPhase::InternalSensibleEnergyTemperatureFunction(const PetscReal *conserved, PetscReal T, PetscReal *internalEnergy, void *ctx) {
+    PetscFunctionBeginUser;
+    // same as internalSensibleEnergyFunction
+    auto functionContext = (FunctionContext *)ctx;
+
+    // get the velocity for kinetic energy
+    PetscReal density = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHO];
+    PetscReal ke = 0.0;
+    for (PetscInt d = 0; d < functionContext->dim; d++){
+        ke += PetscSqr(conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOU + d] / density);
+    }
+    ke *= 0.5;
+
+    // compute internal energy
+    *internalEnergy = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOE] / density - ke;
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::eos::TwoPhase::SensibleEnthalpyFunction(const PetscReal *conserved, PetscReal *sensibleEnthalpy, void *ctx) {
+    PetscFunctionBeginUser;
+    // Total Enthalpy == Sensible Enthalpy = e + p/rho
+    auto functionContext = (FunctionContext *)ctx;
+    PetscReal density = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHO];
+
+    PetscReal sensibleInternalEnergy;
+    InternalSensibleEnergyFunction(conserved, &sensibleInternalEnergy, ctx);
+
+    // get kinetic energy for internal energy calculation
+    PetscReal ke = 0.0;
+    for (PetscInt d = 0; d < functionContext->dim; d++){
+        ke += PetscSqr(conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOU + d] / density);
+    }
+    ke *= 0.5;
+
+    // compute internal energy
+    PetscReal internalEnergy = conserved[functionContext->eulerOffset + ablate::finiteVolume::CompressibleFlowFields::RHOE] / density - ke;
+    // simple decode to compute pressure
+    PetscReal derived[6];
+    PetscReal in[10];
+    in[0] = conserved[functionContext->volumeFractionOffset];
+    in[1] = conserved[functionContext->densityVFOffset];
+    in[2] = density;
+    in[3] = internalEnergy;
+    in[4] = functionContext->parameters.gamma1;
+    in[5] = functionContext->parameters.gamma2;
+    if (functionContext->parameters.p01 == 0 && functionContext->parameters.p02 == 0){
```

Comment:
I would recommend not doing this check everytime.  Instead have a SensibleEnthalpyFunction for each possible interaction that handles it directly.  Then return the specific SensibleEnthalpyFunction for the scenario.  If it is easier to think about you could create three equation of states and have this one just create the one needed.

Reply:
I will need your help on how to change this. Still haven't come up with a way to do this efficiently, since this effects the P, T, h, Cv, Cp, and sound speed functions.

Reply:
I would have separate functions for each, something like:

- SensibleEnthalpyFunctionSimpleGasSimpleGas
- SensibleEnthalpyFunctionSimpleGasStiff
- SensibleEnthalpyFunctionStiffStiff

You could do this for each P, T, h, Cv, Cp.  The question is how much extra work is done in the decode state, would it be more efficient to move the parts of each needed decode into the separate P, T, ... functions?  If all of the decode needs to be done everytime and we should keep the function I would replace the array with indexes with a struct to make it easier to read.